cmd/kubeadm/app/cmd/join.go

```diff
@@ -152,206 +153,258 @@ var (
+		options.externalcfg.Discovery.BootstrapToken = nil
+	} else {
+		if len(options.cfgPath) == 0 && len(args) > 1 {
+			klog.Warningf("[join] WARNING: More than one API server endpoint supplied on command line %v. Using the first one.", args)
+		}
+		options.externalcfg.Discovery.BootstrapToken.APIServerEndpoint = args[0]
+	}
+
+	// if not joining a control plane, unset the ControlPlane object
+	if !options.controlPlane {
+		options.externalcfg.ControlPlane = nil
+	}
+
+	ignorePreflightErrorsSet, err := validation.ValidateIgnorePreflightErrors(options.ignorePreflightErrors)
+	if err != nil {
+		return joinData{}, err
+	}
 
-	if defaultcfg.NodeRegistration.Name == "" {
+	if err = validation.ValidateMixedArguments(cmd.Flags()); err != nil {
+		return joinData{}, err
+	}
+
+	// Either use the config file if specified, or convert public kubeadm API to the internal JoinConfiguration
+	// and validates JoinConfiguration
+	if options.externalcfg.NodeRegistration.Name == "" {
 		klog.V(1).Infoln("[join] found NodeName empty; using OS hostname as NodeName")
 	}
 
-	if defaultcfg.ControlPlane != nil && defaultcfg.ControlPlane.LocalAPIEndpoint.AdvertiseAddress == "" {
+	if options.externalcfg.ControlPlane != nil && options.externalcfg.ControlPlane.LocalAPIEndpoint.AdvertiseAddress == "" {
 		klog.V(1).Infoln("[join] found advertiseAddress empty; using default interface's IP address as advertiseAddress")
 	}
 
-	internalCfg, err := configutil.JoinConfigFileAndDefaultsToInternalConfig(cfgPath, defaultcfg)
+	cfg, err := configutil.JoinConfigFileAndDefaultsToInternalConfig(options.cfgPath, options.externalcfg)
 	if err != nil {
-		return nil, err
+		return joinData{}, err
 	}
 
 	// override node name and CRI socket from the command line options
-	if defaultcfg.NodeRegistration.Name != "" {
-		internalCfg.NodeRegistration.Name = defaultcfg.NodeRegistration.Name
+	if options.externalcfg.NodeRegistration.Name != "" {
+		cfg.NodeRegistration.Name = options.externalcfg.NodeRegistration.Name
 	}
-	if defaultcfg.NodeRegistration.CRISocket != kubeadmapiv1beta1.DefaultCRISocket {
-		internalCfg.NodeRegistration.CRISocket = defaultcfg.NodeRegistration.CRISocket
+	if options.externalcfg.NodeRegistration.CRISocket != kubeadmapiv1beta1.DefaultCRISocket {
+		cfg.NodeRegistration.CRISocket = options.externalcfg.NodeRegistration.CRISocket
 	}
 
-	if internalCfg.ControlPlane != nil {
-		if err := configutil.VerifyAPIServerBindAddress(internalCfg.ControlPlane.LocalAPIEndpoint.AdvertiseAddress); err != nil {
-			return nil, err
+	if cfg.ControlPlane != nil {
+		if err := configutil.VerifyAPIServerBindAddress(cfg.ControlPlane.LocalAPIEndpoint.AdvertiseAddress); err != nil {
+			return joinData{}, err
 		}
 	}
 
+	return joinData{
+		cfg:                   cfg,
+		ignorePreflightErrors: ignorePreflightErrorsSet,
+		outputWriter:          out,
+	}, nil
+}
+
+// Run executes worker node provisioning and tries to join an existing cluster.
+func (j *joinData) Run() error {
```

Comment:
i guess the plan for followup PRs is to create a join phase runner and add the phases to it instead of calling them in this Run() function?

one concern here is that we don't have a clear separation between init and join phases - i.e. the init phases are under this namespace:
https://github.com/kubernetes/kubernetes/tree/8288f9e457561e666ee0ab1dbbd2b14c36fbdaed/cmd/kubeadm/app/cmd/phases

maybe we should put them under `phases/init`, `phases/join`?

Reply:
@neolit123 yes next pr will add the join phase runner.
I don't think we should have separation between phases, instead when possible we should reuse existing phases into the join workflow, but let's see this once we agree upon a list of phases for this workflow